### ISSUE TITLE
`CA2000` fires on null check under complex conditions

### ISSUE THREAD

**Issue by RenderMichael** [dotnet/roslyn-analyzers#7188]
### Analyzer

**Diagnostic ID**: [CA2000](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

.NET 8

### Describe the bug

This only seems to trigger in complex scenarios.

``` csharp
internal async Task<MyClass?> Search(CancellationToken token = default)
{
    await _http.SendAsync(token).ConfigureAwait(false);

    // do not dispose unless result is null - result takes control of the object lifetime
    var response = JsonDocument.Parse("");

    MyClass? result = MatchOrDefault(response); // CA2000
    if (result is null)
    {
        response.Dispose();
        return null;
    }

    return result;
}

internal static MyClass? MatchOrDefault(JsonDocument document)
{
    foreach (JsonElement hit in document.RootElement.EnumerateArray())
    {
        JsonElement fields = hit.GetProperty("_source");

        if (fields.ValueKind is JsonValueKind.Object)
        {
            return new MyClass(document);
        }
    }
    return null;
}

internal class MyClass(JsonDocument document) : IDisposable
{
    public void Dispose() => document.Dispose();
}
```

### Expected behavior

No diagnostic

### Actual behavior

Diagnostic triggered

## Additional context

This reproduction is really finnicky. For example, if the null check `result is null` is changed to `result == null`, the diagnostic goes away. Likewise, the complexity of `MatchOrDefault` is necessary, if that gets simplified the diagnostic goes away as well (I trimmed the repro down as minimally as I could reasonable get it).
